### PR TITLE
fix #2312 add user provided state to Retry

### LINKED ISSUE #2312: Support state on `RetrySignal` / `RetryXxxSpec`
## Motivation
The last gap in functionality between `reactor-extra` 's `reactor.retry.Retry` builder and the `RetrySpec` and `RetryBackoffSpec` is the support of the "application context".

## Desired solution
Provide a _similar_ feature with access to some state via the `RetrySignal` interface.

## Considered alternatives

 * Option 1: expose `retryContext[View]` on `RetrySignal`, defaults to `Context.empty()`. Have methods on the specs to add or replace the retryContext. Have the spec always copy the `RetrySignal` and replace the `retryContext` with that of the spec.
 * Option 2: expose `getContext[View]` on `RetrySignal`. Same as above but the empty base context is replaced with the sequence's own `currentContext()`.
 * Option 3: generify both `Retry` and `RetrySpec` to allow for a single state object `<T>`. It must be mutable and will be exposed mainly through the copy of `ImmutableRetrySignal` done by `RetrySpec`.

## Challenges
 * should this target 3.3.x or 3.4 ?
 * reactor-extra's approach of having the builder typed to the class of the state (and limiting to a single state element) might be harder to achieve, as currently `Retry` is not generified. This is especially harder if targeting 3.3

### PR DESCRIPTION
**Pull request by ericbottard**

### CODE REVIEW

File: reactor-core/src/main/java/reactor/util/retry/Retry.java [hunk 1]
```diff
@@ -60,6 +68,10 @@
 	 */
 	public abstract Publisher<?> generateCompanion(Flux<RetrySignal> retrySignals);
 
+
+	public ContextView retryContext() {return retryContext;}
```
Comment (bsideup) [inline reactor-core/src/main/java/reactor/util/retry/Retry.java:72]:
```suggestion
	public ContextView retryContext() {
		return retryContext;
	}
```

Comment (bsideup) [reply reactor-core/src/main/java/reactor/util/retry/Retry.java:72]: Could you also please clarify `retryContextView()` vs `retryContext()`?

Comment (ericbottard) [reply reactor-core/src/main/java/reactor/util/retry/Retry.java]: I added a comment on `Retry.retryContext()`.
We could also decide to rename `RetrySignal.retryContextView() -> RetrySignal.retryContext()`, let me know what you think.

File: reactor-core/src/main/java/reactor/util/retry/Retry.java [hunk 2]
```diff
@@ -47,6 +49,12 @@
  */
 public abstract class Retry {
 
+	private final ContextView retryContext;
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/util/retry/Retry.java:52]: I was wondering if we should add `retryContext` directly to `Retry` like that. Alternatively we could isolate the feature to `RetrySpec` and `RetryBackoffSpec`, injecting the `ContextView` into the `ImmutableRetrySignal` they expose to the companion publisher they create (since it is already a `.copy()`)...

That said, the idea of the feature leaks into `RetrySignal` interface anyway, so it could be argued that it must be enforced at `Retry` level.

At a minimum, `Retry` no-arg constructor should be maintained though. This adding of a protected ctor introduces a breaking change that we can avoid.

File: reactor-core/src/main/java/reactor/util/retry/ImmutableRetrySignal.java [hunk 3]
```diff
@@ -27,12 +30,18 @@
 	final long      failureTotalIndex;
 	final long      failureSubsequentIndex;
 	final Throwable failure;
+	final ContextView retryContext;
 
 	ImmutableRetrySignal(long failureTotalIndex, long failureSubsequentIndex,
 			Throwable failure) {
+		this(failureTotalIndex, failureSubsequentIndex, failure, Context.empty());
+	}
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/util/retry/ImmutableRetrySignal.java:38]: format: missing newline

File: reactor-core/src/main/java/reactor/core/publisher/FluxRetryWhen.java [hunk 4]
```diff
@@ -113,12 +114,15 @@ public Object scanUnsafe(Attr key) {
...
 				Sinks.Many<Retry.RetrySignal> signaller,
-				CorePublisher<? extends T> source) {
+				CorePublisher<? extends T> source,
+				ContextView retryContext) {
 			super(actual);
 			this.signaller = signaller;
 			this.source = source;
 			this.otherArbiter = new Operators.DeferredSubscription();
 			this.context = actual.currentContext();
+
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/FluxRetryWhen.java:124]: nitpick format: unnecessary newline

**Review by simonbasle — CHANGES_REQUESTED** [review]

File: reactor-core/src/main/java/reactor/util/retry/Retry.java (hunk 2 above)
Comment (ericbottard) [reply reactor-core/src/main/java/reactor/util/retry/Retry.java:52]: > I was wondering if we should add retryContext directly to Retry like that. Alternatively we could isolate the feature to RetrySpec and RetryBackoffSpec, injecting the ContextView into the ImmutableRetrySignal they expose to the companion publisher they create (since it is already a .copy())...

We had that questioning with @aneveu but the issue is this: `Retry` exists so that users can bring their own, and the framework only knows about `Retry` when it creates `RetrySignals`, so the context __has to__ belong to `Retry` (and `RetrySignal`).

> At a minimum, Retry no-arg constructor should be maintained though. This adding of a protected ctor introduces a breaking change that we can avoid.

will do

Comment (ericbottard) [reply reactor-core/src/main/java/reactor/util/retry/Retry.java:52]: > This adding of a protected ctor introduces a breaking change that we can avoid.

BTW, wondering why japicmp did not detect that...

Comment (ericbottard) [reply reactor-core/src/main/java/reactor/util/retry/Retry.java:52]: Oh, but we have Retry on branch 3.3.x but not in 3.3.0.RELEASE, that is why...

File: reactor-core/src/test/java/reactor/core/publisher/FluxRetryWhenTest.java [hunk 5]
```diff
@@ -577,13 +577,43 @@ public void inners() {
...
-				.RetryWhenMainSubscriber<>(actual, signaller, when);
+				.RetryWhenMainSubscriber<>(actual, signaller, when, Context.empty());
 
 		List<Scannable> inners = main.inners().collect(Collectors.toList());
 
 		assertThat(inners).containsExactly((Scannable) signaller, main.otherArbiter);
 	}
 
+	@Test
+	public void retryContextExposedOnRetrySignal() {
```
[the remark this answers is not in the document]
Comment (ericbottard) [reply reactor-core/src/test/java/reactor/core/publisher/FluxRetryWhenTest.java:588]: One could argue that's not the spirit of the test: given the discussion above about the fact that the context is borne by `Retry`, this works by design. 

Adding a test about backoff would only test that the context is correctly copied over when builder methods are used (but then this is a different kind of test, with _many_ combinations that we don't have for other fields either).

WDYT?

File: reactor-core/src/main/java/reactor/util/retry/Retry.java (hunk 2 above)
Comment (simonbasle) [reply reactor-core/src/main/java/reactor/util/retry/Retry.java]: > Oh, but we have Retry on branch 3.3.x but not in 3.3.0.RELEASE, that is why...

Ah damn!

> The framework only knows about Retry when it creates RetrySignals, so the context has to belong to Retry (and RetrySignal).

Yeah I saw that... `RetryContext` is a bit different from all the other spec parameters, in the sense that it affects the `RetrySignal`, so yeah it must mean it has to live in `Retry` directly.

File: reactor-core/src/main/java/reactor/util/retry/Retry.java [hunk 6]
```diff
@@ -60,6 +72,16 @@
 	 */
 	public abstract Publisher<?> generateCompanion(Flux<RetrySignal> retrySignals);
 
+	/**
+	 * Return the user provided context that was set at construction time.
+	 *
+	 * @return the user provided context that will be accessible via {@link RetrySignal#retryContextView()}.
+	 */
+	public ContextView retryContext() {
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/util/retry/Retry.java:80]: I'm still unsure about that one. should it be `retryContextView()`? or even simpler, `context()`? (as `Retry` implementors are less likely to implement stuff that is unrelated to retries, unlike eg. `RetrySignal`). In the end, I think I like the simplicity of `retryContext()` though...

File: reactor-core/src/main/java/reactor/util/retry/Retry.java [hunk 7]
```diff
@@ -92,6 +114,16 @@
 		 */
 		Throwable failure();
 
+		/**
+		 * Return a read-only view of the user provided context, which may be used to store
+		 * objects to be reset/rollbacked before a retry.
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/util/retry/Retry.java:119]: I would say "reset/rollbacked or otherwise mutated before or after a retry"

File: reactor-core/src/main/java/reactor/util/retry/Retry.java [hunk 8]
```diff
@@ -92,6 +114,16 @@
...
+		 *
+		 * @return a read-only view of a user provided context.
+		 */
+		default ContextView retryContextView() {
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/util/retry/Retry.java:123]: I agree on using `ContextView` suffix, as this is the part of the API that will be most visible to users and thus further clarifies that this is read only.

File: reactor-core/src/test/java/reactor/core/publisher/FluxRetryWhenTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (simonbasle) [reply reactor-core/src/test/java/reactor/core/publisher/FluxRetryWhenTest.java:588]: ok, maybe not duplicating the full test. I would maybe just test both constructors to ensure the retry context is passed to super ctor in both flavors of specs, wdyt?

**Review by simonbasle — CHANGES_REQUESTED** [review]

File: reactor-core/src/test/java/reactor/core/publisher/FluxRetryWhenTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (ericbottard) [reply reactor-core/src/test/java/reactor/core/publisher/FluxRetryWhenTest.java:588]: Have a look at ff3aed1a and let me know what you think

**Review by simonbasle — APPROVED** [review]

File: reactor-core/src/test/java/reactor/core/publisher/FluxRetryWhenTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (simonbasle) [reply reactor-core/src/test/java/reactor/core/publisher/FluxRetryWhenTest.java:588]: 👍
